fgpyo/fasta/builder.py

```diff
@@ -0,0 +1,231 @@
+    defaults are provided if 'assembly' an/or 'species' are not provided.
+
+    Attributes:
+        name: Unique contig ID, ie., "chr10"
+        assembly: Assembly information, if None default is 'testassembly'
+        species: Species information, if None default is 'testspecies'
+        bases:  The bases to be added to the contig ex "A"
+
+    """
+
+    def __init__(
+        self,
+        name: str,
+        assembly: str = "testassembly",
+        species: str = "testspecies",
+    ):
+        self.name = name
+        self.assembly = assembly
+        self.species = species
+        self.bases = ""
+
+    def add(self, bases: str, times: int) -> "ContigBuilder":
+        """
+        Method for adding bases to a new or existing instance of ContigBuilder.
+
+        Args:
+            bases: The bases to be added to the contig
+            times: The number of times the bases should be repeated
+
+        Example
+        add("AAA", 2) results in the following bases -> "AAAAAA"
+        """
+        # Remove any spaces in string and enforce upper case format
+        bases = bases.replace(" ", "").upper()
+        self.bases += str(bases * times)
+        return self
+
+
+class FastaBuilder:
+    """Builder for constructing sets of one or more contigs.
+
+    Provides the ability to manufacture sets of contigs from minimal input, and automatically
+    generates the information necessary for writing the FASAT file, index, and dictionary.
+
+    A builder is constructed from an assembly, species, and line length. All attributes have
+    defaults, however these can be overwritten.
+
+    Contigs are added to FastaBuilder using:
+    :func:`~fgpyo.fasta.builder.FastaBuilder.add`
+
+    Bases are added to existing contigs using:
+    :func:`~fgpyo.fasta.builder.FastaBuilder.add.add`
+
+    Once accumulated the contigs can be written to a file using:
+    :func:`~fgpyo.fasta.builder.FastaBuilder.to_file`
+
+    Calling to_file() will also generate the fasta index (.fai) and sequence dictionary (.dict).
+
+    Attributes:
+        assembly: Assembly information, if None default is 'testassembly'
+        species: Species, if None default is 'testspecies'
+        line_length: Desired line length, if None default is 80
+        contig_builders: Private dictionary of contig names and instances of ContigBuilder
+    """
+
+    def __init__(
+        self,
+        assembly: str = "testassembly",
+        species: str = "testspecies",
+        line_length: int = 80,
+    ):
+        self.assembly: str = assembly
+        self.species: str = species
+        self.line_length: int = line_length
+        self.__contig_builders: Dict[str, ContigBuilder] = {}
+
+    def __getitem__(self, key: str) -> ContigBuilder:
+        """ Access instance of ContigBuilder by name """
+        return self.__contig_builders[key]
+
+    def add(
+        self,
+        name: str,
+        assembly: str = "testassembly",
+        species: str = "testspecies",
```

Comment:
Ok, so here's what I was trying to say before.  I think these two parameters should be specified as:

```python
assembly: Optional[str] = None,
species: Optional[str] = None,
```

And then below, before passing these to `ContigBuilder()` you'll do:

```
assembly = assembly if assembly is not None else self.assembly
species = species if species is not None else self.species
```

So that the default is to use whatever value is held on the `FastaBuilder` unless a value is specified by the user.

Reply:
Got it! Thank you for the clarification. I have changed `assembly` and `species` attributes in `ContigBuilder` to be required strings with no default. While the parameters `assembly` and `species` in `FastaBuilder.add()` are Optional as shown above with defaults for `ContigBuilder.assembly/species` being the `FastaBuilder` attributes. See changes in commit [d31a292](https://github.com/fulcrumgenomics/fgpyo/commit/d31a2926a18cf995d2bd780a40d5bcc9f31b6989)